runtime/compare.c

```diff
@@ -276,33 +304,57 @@ static intnat do_compare_val(struct compare_stack* stk,
       /* Compare sizes first for speed */
       if (sz1 != sz2) return sz1 - sz2;
       if (sz1 == 0) break;
-      /* Remember that we still have to compare fields 1 ... sz - 1 */
-      if (sz1 > 1) {
-        sp++;
-        if (sp >= stk->limit) sp = compare_resize_stack(stk, sp);
-        sp->v1 = &Field(v1, 1);
-        sp->v2 = &Field(v2, 1);
-        sp->count = sz1 - 1;
+      /* Remember that we still have to compare fields 1 ... sz - 1. */
+      if (sz1 > 1 || CAMLunlikely(signal_poll_timer == 0)) {
```

Comment:
These `signal_poll_timer == 0` comparisons are fragile.  If `signal_poll_timer` can be decremented several times before testing, you can miss the `0` value and get a negative `signal_poll_timer`, which will test negative for 2^31 more iterations or so.  This can happen in the `Forward_tag` case, which `continue`s straight to the beginning of the `while (1)` loop, decrementing `signal_poll_timer` a second time.  Please write robust tests `signal_poll_timer <= 0`.

Reply:
Done, thanks.

(I thought about changing those `continue` to `goto continue_after_tick` to avoid this, because using `== 0` is neater, but I went for the more robust test.)

Reply:
If you want a `signal_poll_timer == 0` test,  decrement `signal_poll_timer` just before the test, not two pages above, so that it is obvious that the timer cannot skip value 0.

Reply:
Yes, something like `if (--poll_timer == 0) must_poll = 1;` as the tick instruction, and then checking the `must_poll` variable as the poll test. But I am content with `<= 0`.